lib/interfaces/include/BMSDriverGroup.h

```diff
@@ -0,0 +1,332 @@
+     * bytes are communicated to the device correctly.
+     * This code was taken directly from the data sheet, page 76
+     * @post Instantiate the public pec15Table array of size 256. This variable should never be modified afterwards
+     */
+    void generate_PEC_table();
+
+    /**
+     * WAKEUP:
+     * The Wakeup protocol will move the device from the SLEEP to STANDBY / REFUP state
+     * Reference pages 51-52 on the data sheet for specific information.
+     * @pre we don't care; for us, it's good to perform a wakeup to guarantee proper data propogation
+     * @post essentially sets connected pin to LOW for period of time, then to HIGH for period of time
+     */
+    void start_wakeup_protocol();
+
+    /**
+     * Returns the delay time depending on what state we are in: SLEEP, IDLE, or MEASURE
+     * @return int number of milliseconds
+     */
+    int get_delay_time();
+
+    /* -------------------- READING DATA FUNCTIONS -------------------- */
+
+    /**
+     * Reads voltage data from ALL 4 Cell Voltage Registers on the device.
+     * A register, for those that don't know (I didn't), is memory.
+     * @pre in order to actually "read" anything, we need to call wakeup() and send data over SPI
+     * @post store all 126 cell voltages into the cell_voltages container
+     * AND record the max / minimum values and locations of them
+     */
+    void read_voltages();
+
+    /**
+     * Reads GPIO data from ALL 2 Auxillary GPIO Registers on the device.
+     * For each BMS segment, there is 6 board thermistors and 2 humidity sensors.
+     * NOTE: Conversions are different depending on which we are reading.
+     * @pre in order to actually "read" anything, we need to call wakeup() and send data over SPI
+     * @post store all temperature data into the board_temperatures container
+     * AND record the maximum value and locations
+     */
+    void read_GPIOs();
+
+    /* -------------------- WRITING DATA FUNCTIONS -------------------- */
+
+    /**
+     * Writes the device configuration
+     * @pre needs access to undervoltage, overvoltage, configuration MACROS, and discharge data
+     * @post sends packaged data over SPI
+     */
+    void write_configuration(uint8_t dcto_mode);
+
+    /**
+     * Writes command to start cell voltage ADC converion
+     * @post packaged data transferred over SPI, need to delay before we can read
+     */
+    void start_cell_voltage_ADC_conversion();
+
+    /**
+     * Writes command to start GPIO ADC conversion
+     * @post packaged data transfered over SPI, need to delay before we can read
+     * This means that we have to call a delay before continuing.
+     */
+    void start_GPIO_ADC_conversion();
+
+    /* -------------------- PRINT DATA FUNCTIONS -------------------- */
+
+    /**
+     * Prints all the cell voltage information
+     * Every cell voltage, the minimum, maximum, and locations of those
+     */
+    void print_voltage_data();
```

Comment:
no need to have this function here. if the user wants to print the voltages, the driver should not implement this, they should use the driver to get the information to print.